transaction-pool/src/transactions.rs

```diff
@@ -0,0 +1,213 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{fmt, mem};
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use smallvec::SmallVec;
+
+use ready::{Ready, Readiness};
+use scoring::{self, Scoring};
+
+#[derive(Debug)]
+pub enum AddResult<T> {
+	Ok(Arc<T>),
+	TooCheapToEnter(T),
+	TooCheap {
+		old: Arc<T>,
+		new: T,
+	},
+	Replaced {
+		old: Arc<T>,
+		new: Arc<T>,
+	},
+	PushedOut {
+		old: Arc<T>,
+		new: Arc<T>,
+	},
+}
+
+/// Represents all transactions from a particular sender ordered by nonce.
+const PER_SENDER: usize = 8;
+#[derive(Debug)]
+pub struct Transactions<T, S: Scoring<T>> {
+	// TODO [ToDr] Consider using something that doesn't require shifting all records.
+	transactions: SmallVec<[Arc<T>; PER_SENDER]>,
+	scores: SmallVec<[S::Score; PER_SENDER]>,
+	_score: PhantomData<S>,
+}
+
+impl<T, S: Scoring<T>> Default for Transactions<T, S> {
+	fn default() -> Self {
+		Transactions {
+			transactions: Default::default(),
+			scores: Default::default(),
+			_score: PhantomData,
+		}
+	}
+}
+
+impl<T: fmt::Debug, S: Scoring<T>> Transactions<T, S> {
+	pub fn is_empty(&self) -> bool {
+		self.transactions.is_empty()
+	}
+
+	pub fn len(&self) -> usize {
+		self.transactions.len()
+	}
+
+	pub fn iter(&self) -> ::std::slice::Iter<Arc<T>> {
+		self.transactions.iter()
+	}
+
+	pub fn worst_and_best(&self) -> Option<((S::Score, Arc<T>), (S::Score, Arc<T>))> {
+		let len = self.scores.len();
+		self.scores.get(0).cloned().map(|best| {
+			let worst = self.scores[len - 1].clone();
+			let best_tx = self.transactions[0].clone();
+			let worst_tx = self.transactions[len - 1].clone();
+
+			((worst, worst_tx), (best, best_tx))
+		})
+	}
+
+	pub fn find_next(&self, tx: &T, scoring: &S) -> Option<(S::Score, Arc<T>)> {
+		self.transactions.binary_search_by(|old| scoring.compare(old, &tx)).ok().and_then(|index| {
+			let index = index + 1;
+			if index >= self.scores.len() {
+				None
+			} else {
+				Some((self.scores[index].clone(), self.transactions[index].clone()))
+			}
+		})
+	}
+
+	pub fn add(&mut self, tx: T, scoring: &S, max_count: usize) -> AddResult<T> {
+		let index = match self.transactions.binary_search_by(|old| scoring.compare(old, &tx)) {
+			Ok(index) => index,
+			Err(index) => index,
+		};
+
+		// Insert at the end.
+		let len = self.transactions.len();
+		if index == len {
+			// trying to insert least worth transaction
+			return if len == max_count {
+				AddResult::TooCheapToEnter(tx)
+			} else {
+				let shared = Arc::new(tx);
+				self.transactions.push(shared.clone());
+				self.scores.push(Default::default());
+				scoring.update_scores(&self.transactions, &mut self.scores, scoring::Change::InsertedAt(index));
+
+				AddResult::Ok(shared)
+			}
+		}
+
+		// Decide if the transaction should be replaced
+		match scoring.choose(&self.transactions[index], &tx) {
```

Comment:
If I were you, I'd introduce `result` value here to put an emphasis on a fact that we return a value from that large `match`.